File: src/main/org/nlogo/prim/gui/_movieclose.java
package org.nlogo.prim.gui;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Syntax;
import org.nlogo.window.GUIWorkspace;

public final strictfp class _movieclose
    extends org.nlogo.nvm.Command {
  @Override
  public Syntax syntax() {
    return Syntax.commandSyntax();
  }

  @Override
  public void perform(final org.nlogo.nvm.Context context)
      throws LogoException {
    if (!(workspace instanceof GUIWorkspace)) {
      throw new EngineException(
          context, this, token().name() + " can only be used in the GUI");
    }
    workspace.waitFor
        (new org.nlogo.api.CommandRunnable() {
          public void run() throws LogoException {
            if (((GUIWorkspace) workspace).movieEncoder == null) {
              throw new EngineException
                  (context, _movieclose.this,
                      "Must call MOVIE-START first");
            }

            new org.nlogo.swing.ModalProgressTask
                (((GUIWorkspace) workspace).getFrame(),
                    new Runnable() {
                      public void run() {
                        ((GUIWorkspace) workspace)
                            .movieEncoder.stop();
                        ((GUIWorkspace) workspace).movieEncoder = null;
                      }
                    },
                    "Exporting movie...");
          }
        });
    context.ip = next;
  }
}

File: src/main/org/nlogo/prim/gui/_movieclose.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.gui

import org.nlogo.api.{ CommandRunnable, Syntax }
import org.nlogo.nvm.{ Command, Context, EngineException }
import org.nlogo.window.GUIWorkspace

class _movieclose extends Command {
  override def syntax =
    Syntax.commandSyntax
  override def perform(context: Context) {
    workspace match {
      case gw: GUIWorkspace =>
        workspace.waitFor(
          new CommandRunnable {
            override def run() {
              if (gw.movieEncoder == null)
                throw new EngineException(context, _movieclose.this,
                      "Must call MOVIE-START first");
              org.nlogo.swing.ModalProgressTask.apply(
                gw.getFrame, "Exporting movie...",
                new Runnable() {
                  override def run() {
                    gw.movieEncoder.stop()
                    gw.movieEncoder = null
                  }})}})
      case _ =>
        throw new EngineException(
          context, this, token.name + " can only be used in the GUI")
    }
    context.ip = next
  }
}
